### ISSUE TITLE
 Not found: Key <variable_name> not found in checkpoint even though it exists in meta graph

### ISSUE THREAD

**Issue by kevinlu1211** [tensorflow/tensorflow#11597]
### System information
Python version: 3.6
Tensorflow version: 1.1

Here are the commands needed to reproduce the error:

```
import tensorflow as tf
import numpy as np

    
tf.reset_default_graph()

# Create 100 phony x, y data points in NumPy, y = x * 0.1 + 0.3
x_data = np.random.rand(100).astype(np.float32)
y_data = x_data * 0.1 + 0.3

# Try to find values for W and b that compute y_data = W * x_data + b
# (We know that W should be 0.1 and b 0.3, but TensorFlow will
# figure that out for us.)
W = tf.Variable(tf.random_uniform([1], -1.0, 1.0), name='W')
b = tf.Variable(tf.zeros([1]), name='b')
y = W * x_data + b

# Minimize the mean squared errors.
loss = tf.reduce_mean(tf.square(y - y_data))
optimizer = tf.train.GradientDescentOptimizer(0.5)
opt_op = optimizer.minimize(loss)

# Track the moving averages of all trainable variables.
ema = tf.train.ExponentialMovingAverage(decay=0.9999)
variables = tf.trainable_variables()
print(variables)
averages_op = ema.apply(tf.trainable_variables())
with tf.control_dependencies([opt_op]):
    train_op = tf.group(averages_op)

# Before starting, initialize the variables.  We will 'run' this first.
init = tf.global_variables_initializer()

saver = tf.train.Saver(tf.trainable_variables())

# Launch the graph.
sess = tf.Session()
sess.run(init)

# Fit the line.
for _ in range(201):
    sess.run(train_op)

w_reference = sess.run('W/ExponentialMovingAverage:0')
b_reference = sess.run('b/ExponentialMovingAverage:0')

saver.save(sess, os.path.join("model_ex1"))

tf.reset_default_graph()

tf.train.import_meta_graph("model_ex1.meta")
sess = tf.Session()

print('------------------------------------------------------')
for var in tf.global_variables():
    print('all variables: ' + var.op.name)
for var in tf.trainable_variables():
    print('normal variable: ' + var.op.name)
for var in tf.moving_average_variables():
    print('ema variable: ' + var.op.name)
print('------------------------------------------------------')


restore_vars = {}

ema = tf.train.ExponentialMovingAverage(1.0)
for var in tf.trainable_variables():
    print('%s: %s' % (ema.average_name(var), var.op.name))
    restore_vars[ema.average_name(var)] = var

saver = tf.train.Saver(restore_vars, name='ema_restore')

saver.restore(sess, os.path.join("model_ex1"))

w_restored = sess.run('W:0')
b_restored = sess.run('b:0')

print(w_reference)
print(w_restored)
print(b_reference)
print(b_restored)

```

**Comment by printdhruv** [comment]
Can you explain what you are trying to do in following lines?
```
saver = tf.train.Saver(restore_vars, name='ema_restore')
saver.restore(sess, os.path.join("model_ex1"))
```

**Comment by kevinlu1211** [comment]
I'm telling the saver to load the EMA weights instead of the normal weights, or have I done it wrong?

**Comment by printdhruv** [comment]
I think tf.train.saver() must be saved. Passing arguments to the constructor won't be sufficient I guess. Please refer, [https://www.tensorflow.org/api_docs/python/tf/train/Saver](https://www.tensorflow.org/api_docs/python/tf/train/Saver)

**Comment by kevinlu1211** [comment]
What do you mean by that? Haven't I already used saver.save after doing gradient descent?

```
# Fit the line.
for _ in range(201):
    sess.run(train_op)

w_reference = sess.run('W/ExponentialMovingAverage:0')
b_reference = sess.run('b/ExponentialMovingAverage:0')

saver.save(sess, os.path.join("model_ex1"))
```

**Comment by printdhruv** [comment]
I am referring `tf.train.Saver(restore_vars, name='ema_restore')`

**Comment by kevinlu1211** [comment]
I've just looked at the docs, and I can't seem to find the code that says to save tf.train.saver() before loading it. If you got it working could you show me your code? Thanks

**Comment by reedwm** [comment]
This question is better asked on  [StackOverflow](http://stackoverflow.com/questions/tagged/tensorflow) since it is not a  bug or feature request. There is also a larger community that reads questions there. Thanks!

That said, when you recreate the `ExponentialMovingAverage` after saving the checkpoint, it is unaware of the current variables and shadow variables, since you just created it. Then, when you call `ema.average_name(var)`, it recreates new shadow variables since it doesn't know about the old shadow variables. I'm not sure how to solve this while still using a `MetaGraph`; try asking StackOverflow or rebuild the model instead of using a `MetaGraph`.

**Comment by Rajamohanreddyai** [comment]
Hello all,

just follow the below video and export your own model with in a 10 seconds

https://youtu.be/w0Ebsbz7HYA
